### PR TITLE
Display an error during failure of fallback UIA

### PR DESCRIPTION
**Pull request by govynnus**
Signed-off-by: Callum Brown <callum@calcuode.com>

See discussion in https://github.com/matrix-org/synapse/pull/10142#discussion_r661613726
This doesn't fix anything existing because none of the current fallbacks really fail, but is required for decent UX with the registration token fallback in #10142

### CODE REVIEW

File: synapse/handlers/auth.py [hunk 1]
```diff
@@ -627,23 +627,28 @@ async def check_ui_auth(
 
     async def add_oob_auth(
         self, stagetype: str, authdict: Dict[str, Any], clientip: str
-    ) -> bool:
+    ) -> Any:
```
Comment (clokep) [inline synapse/handlers/auth.py:630]: It looks like the result of this isn't even used anymore after this?

If it is, I'd like to see a more specific type here than `Any`. If it isn't, we should have it return `None` and properly document what it raises.

Comment (govynnus) [reply synapse/handlers/auth.py:630]: Yep, I'll do that

Comment (govynnus) [reply synapse/handlers/auth.py]: How's that?

File: synapse/handlers/auth.py [hunk 2]
```diff
@@ -627,23 +627,28 @@ async def check_ui_auth(
...
-            raise LoginError(400, "", Codes.MISSING_PARAM)
+            raise LoginError(
+                400, f"Unknown UIA stage type: {stagetype}", Codes.INVALID_PARAM
+            )
         if "session" not in authdict:
-            raise LoginError(400, "", Codes.MISSING_PARAM)
+            raise LoginError(400, "Missing session ID", Codes.MISSING_PARAM)
 
+        # If authentication fails a LoginError is raised, so if we do get a
+        # result we know authentication has succeeded.
```
Comment (clokep) [inline synapse/handlers/auth.py:647]:
```suggestion
        # If authentication fails a LoginError is raised. Otherwise, store
        # the successful result.
```

File: synapse/handlers/ui_auth/checkers.py [hunk 3]
```diff
@@ -49,7 +49,7 @@ async def check_auth(self, authdict: dict, clientip: str) -> Any:
             clientip: The IP address of the client.
 
         Raises:
-            SynapseError if authentication failed
+            LoginError if authentication failed.
 
         Returns:
             The result of authentication (to pass back to the client?)
```
Comment (clokep) [inline synapse/handlers/ui_auth/checkers.py:55]: I don't suppose we can clarify this comment while we're looking at all of this!

File: synapse/rest/client/v2_alpha/auth.py [hunk 4]
```diff
@@ -95,29 +95,32 @@ async def on_POST(self, request, stagetype):
...
+                    LoginType.RECAPTCHA, authdict, request.getClientIP()
+                )
+            except LoginError as e:
+                # Authentication failed, let user try again
                 html = self.recaptcha_template.render(
                     session=session,
                     myurl="%s/r0/auth/%s/fallback/web"
                     % (CLIENT_API_PREFIX, LoginType.RECAPTCHA),
                     sitekey=self.hs.config.recaptcha_public_key,
+                    error=e.msg,
```
Comment (clokep) [inline synapse/rest/client/v2_alpha/auth.py:109]: Are we sure that these error messages are suitable for displaying to the user? Looking through the checkers I see a few that raise `LoginError` with a blank message.

**Review by clokep — CHANGES_REQUESTED** [review]
Looks better overall! Anything we can do to make this code clearer is great!

**Comment by clokep** [comment]
>  This doesn't fix anything existing because none of the current fallbacks really fail, but is required for decent UX with the registration token fallback in #10142

I'm not sure I agree. It looks like a broken client (sending the wrong parameters) or if the recaptcha endpoint was down could cause ugly errors to pop back into the user's face. Is that what this is improving?

Could you include a screenshot of how the error renders, by the way?

File: synapse/handlers/ui_auth/checkers.py (hunk 3 above)
Comment (govynnus) [reply synapse/handlers/ui_auth/checkers.py:55]: Clarify in what way? Say why/how LoginError is used?

Comment (clokep) [reply synapse/handlers/ui_auth/checkers.py:55]: > (to pass back to the client?)

It seems whoever wrote this comment wasn't sure why we care about this. If you've groked this and know the answer it would be great to say *why* we care about the result of the authentication.

Comment (govynnus) [reply synapse/handlers/ui_auth/checkers.py:55]: The only checkers that return something other than just `True` are the threepid ones:
https://github.com/matrix-org/synapse/blob/8c654b73095a594b36101aa81cf91a8e1bebc29f/synapse/handlers/ui_auth/checkers.py#L206

And the only place I can see the result is accessed is here, ~~but I have no clue where the user ID would get into the result field of a UIA stage.~~ It looks like the [password provider docs](https://matrix-org.github.io/synapse/latest/password_auth_providers.html#optional-methods) (which you wrote?) mention that 
https://github.com/matrix-org/synapse/blob/8c654b73095a594b36101aa81cf91a8e1bebc29f/synapse/handlers/auth.py#L377-L405

Where `result` comes from the only ever use of `get_completed_ui_auth_stages`:
https://github.com/matrix-org/synapse/blob/8c654b73095a594b36101aa81cf91a8e1bebc29f/synapse/handlers/auth.py#L605

There's also a section in the SSO handler which sets the result field, but I'm not sure if that's ever looked at later:
https://github.com/matrix-org/synapse/blob/8c654b73095a594b36101aa81cf91a8e1bebc29f/synapse/handlers/sso.py#L690-L704

I am currently using the fact that the result is stored for expiring registration tokens, but I think I could use the UIA session data instead.
https://github.com/govynnus/synapse/blob/e9435f8f6c9ba1497f689e3963c9c005c1d71dfc/synapse/handlers/ui_auth/checkers.py#L294
https://github.com/govynnus/synapse/blob/e9435f8f6c9ba1497f689e3963c9c005c1d71dfc/synapse/storage/databases/main/ui_auth.py#L333-L373

So basically no, I don't grok it.
Although I'm pretty sure it's not passed back to the client :-)

Comment (clokep) [reply synapse/handlers/ui_auth/checkers.py:55]: Hah, thanks for taking another look! We can just leave it as is then!

**Comment by clokep** [comment]
@govynnus Is this ready for another round of review? (Also please take a look at https://github.com/matrix-org/synapse/pull/10561#issuecomment-896770861)

File: synapse/rest/client/v2_alpha/auth.py (hunk 4 above)
Comment (govynnus) [reply synapse/rest/client/v2_alpha/auth.py:109]: Yeah you're right, the errors are more for devs and people troubleshooting than users I think. From what I can tell the errors would be caused by a broken client rather than the user.

Comment (clokep) [reply synapse/rest/client/v2_alpha/auth.py:109]: I think we should at least give some string back for those cases so that people can ask a server admin about it. Right now some of them will end up just saying `Error: `.

Comment (govynnus) [reply synapse/rest/client/v2_alpha/auth.py:109]: I added a message to the couple LoginErrors that were missing one https://github.com/matrix-org/synapse/pull/10561/commits/8d6a469a43e276a40862eb8a1f2993717572320e

File: docs/upgrade.md [hunk 5]
```diff
@@ -112,6 +112,16 @@ environment variable.
 See [using a forward proxy with Synapse documentation](setup/forward_proxy.md) for
 details.
 
+## UIA fallback templates can display errors
```
Comment (clokep) [inline docs/upgrade.md:115]: Let's be explicit:

```suggestion
## User-interactive authentication fallback templates can now display errors
```

File: changelog.d/10561.bugfix [hunk 6]
```diff
@@ -0,0 +1 @@
+Handle failure of User-Interactive Authentication when using a fallback. Contributed by Callum Brown.
```
Comment (clokep) [inline changelog.d/10561.bugfix:1]: Is this handling the failure or displaying an error? It seems it was previously "handled", but by displaying a JSON error message to the user?

File: docs/upgrade.md [hunk 7]
```diff
@@ -112,6 +112,16 @@ environment variable.
...
+
+This may affect you if you make use of custom HTML templates for the
+[reCAPTCHA](../synapse/res/templates/recaptcha.html) and
+[terms](../synapse/res/templates/terms.html) fallback pages.
+There is no configuration option for these templates.
+
+The template can now use an optional `error` variable, which is defined if
+the authentication process failed. See the default templates linked above for
+an example.
```
Comment (clokep) [inline docs/upgrade.md:124]:
```suggestion
The template is now provided an `error` variable if the authentication 
process failed. See the default templates linked above for an example.
```

File: docs/upgrade.md [hunk 8]
```diff
@@ -112,6 +112,16 @@ environment variable.
 See [using a forward proxy with Synapse documentation](setup/forward_proxy.md) for
 details.
 
+## UIA fallback templates can display errors
+
+This may affect you if you make use of custom HTML templates for the
+[reCAPTCHA](../synapse/res/templates/recaptcha.html) and
+[terms](../synapse/res/templates/terms.html) fallback pages.
```
Comment (clokep) [inline docs/upgrade.md:119]:
```suggestion
This may affect you if you make use of custom HTML templates for the
[reCAPTCHA](../synapse/res/templates/recaptcha.html) or
[terms](../synapse/res/templates/terms.html) fallback pages.
```

File: docs/upgrade.md [hunk 9]
```diff
@@ -112,6 +112,16 @@ environment variable.
...
+There is no configuration option for these templates.
```
Comment (clokep) [inline docs/upgrade.md:120]: I'm not sure what this means -- what would the configuration option do?

**Comment by clokep** [comment]
Looks close! I left some comments on the upgrade notes, but the code is solid.

File: docs/upgrade.md (hunk 9 above)
Comment (govynnus) [reply docs/upgrade.md:120]: Well for other templates (e.g. account validity) there are configuration options so you can give a path to custom templates. I'll just remove that line though because it's probably just confusing.

**Comment by govynnus** [comment]
> Looks close! I left some comments on the upgrade notes, but the code is solid.

Thanks, I fixed the upgrade notes in 70e653e0827ee563420e57c9f821c2a397da950f

**Review by clokep — APPROVED** [review]
